### ISSUE TITLE
Incompleteness in preferring subst-relate candidate in alias-relate goals

### ISSUE THREAD

**Issue by compiler-errors** [rust-lang/trait-system-refactor-initiative#26]
There are 3 ways to prove `AliasRelate` goals: `lhs normalize-to rhs`, `rhs normalize-to lhs`, and `substs-relate`.

If merging the three options via `fn try_merge_responses` fails, we incompletely prefer `substs-relate` over normalizing. https://github.com/rust-lang/rust/blob/a482149598f5aacf3837eee87026dd634f08641c/compiler/rustc_trait_selection/src/solve/alias_relate.rs#L84-L86

It seems like this incompleteness is not necessary if we go with https://github.com/rust-lang/trait-system-refactor-initiative/issues/25 instead.

---

*edit: original description*

ill write this tomorrow since it's 21:50 rn and i want to sleep

tl;dr here's the list of *failing* tests before and after -- red (left-hand side) tests are those that are *fixed* by the incompleteness, and green (right-hand side) are those that begin to fail after the incompleteness: https://www.diffchecker.com/NdEJkqfr/

**Comment by compiler-errors** [comment]
if we were to change this incompleteness to apply only if the response has no infer/opaques, this is the list of failing tests for each scenario: https://www.diffchecker.com/05XWB4w0/

this suggests that we may be able to get away (for now) with only preferrring a subst-relate candidate if it `has_only_region_constraints`.

**Comment by lcnr** [comment]
This is used to fix #2

**Comment by compiler-errors** [comment]
We don't do this anymore as of rust-lang/rust#113901 -- bidirectional normalizes-to is both more conservative and is sufficient in most cases, and #2 was also fixed rust-lang/rust#112399.
